src/plugins/media/media.js

```diff
@@ -0,0 +1,248 @@
+    };
+
+    removeMediaClasses = function() {
+        var type, types;
+        types = [ "video", "audio" ];
+        for ( type in types ) {
+            document.body.classList.remove( "impress-media-" + types[ type ] + "-playing" );
+            document.body.classList.remove( "impress-media-" + types[ type ] + "-paused" );
+        }
+    };
+
+    enhanceMediaNodes = function() {
+        var i, id, media, mediaElement, type;
+
+        media = root.querySelectorAll( "audio, video" );
+        for ( i = 0; i < media.length; i += 1 ) {
+            type = media[ i ].nodeName.toLowerCase();
+
+            // Set an id to identify each media node - used e.g. for cross references by
+            // the consoleMedia plugin
+            mediaElement = media[ i ];
+            id = mediaElement.getAttribute( "id" );
+            if ( id === undefined || id === null ) {
+                mediaElement.setAttribute( "id", "media-" + type + "-" + i );
+                attributeTracker.push( { "node": mediaElement, "attr": "id" } );
+            }
+            gc.addEventListener( mediaElement, "play", onPlay );
+            gc.addEventListener( mediaElement, "playing", onPlay );
+            gc.addEventListener( mediaElement, "pause", onPause );
+            gc.addEventListener( mediaElement, "ended", onEnded );
+        }
+    };
+
+    enhanceMedia = function() {
+        var steps, stepElement, i;
+        enhanceMediaNodes();
+        steps = document.getElementsByClassName( "step" );
+        for ( i = 0; i < steps.length; i += 1 ) {
+            stepElement = steps[ i ];
+
+            gc.addEventListener( stepElement, "impress:stepenter", onStepenter );
+            gc.addEventListener( stepElement, "impress:stepleave", onStepleave );
+        }
+    };
+
+    onStepenterDetectImpressConsole = function() {
+        return {
+            "preview": ( window.frameElement !== null && window.frameElement.id === "preView" ),
+            "slideView": ( window.frameElement !== null && window.frameElement.id === "slideView" )
+        };
+    };
+
+    onStepenter = function( event ) {
+        var stepElement, media, mediaElement, i, onConsole, autoplay;
+        if ( ( !event ) || ( !event.target ) ) {
+            return;
+        }
+
+        stepElement = event.target;
+        removeMediaClasses();
+
+        media = stepElement.querySelectorAll( "audio, video" );
+        for ( i = 0; i < media.length; i += 1 ) {
+            mediaElement = media[ i ];
+
+            // Autoplay when (maybe inherited) autoplay setting is true,
+            // but only if not on preview of the next step in impressConsole
+            onConsole = onStepenterDetectImpressConsole();
+            autoplay = getMediaAttribute( "autoplay", [ mediaElement, stepElement, root ] );
+            if ( autoplay && !onConsole.preview ) {
+                if ( onConsole.slideView ) {
+                    mediaElement.muted = true;
+                }
```

Comment:
Lines 210-212 can be moved inside onStepenterDetectImpressConsole(). (Line 209 can stay as is.)

Reply:
I am not sure about this. That would mute media at the console view even when it is not played automatically. 

I don’t think we should mute every media inside console view even when it maybe is not started. Chances are high, that somebody has an audio on a step and starts them inside the console. Then can not be heard at all, because it is (surprisingly) muted. Audio has not necessarily be played on both screens (or even the main) in contradiction to video that should always be played on both screens, at least inside the presentation view.

I think this should only be done by an additional plugin that handles events between main and console view (we already talked about that). I already had a script working for bi-directional synchronization of media, even avoiding circular propagation from one window to the other, but thought it best to implement this as separate plugin as it assumes the existence of the impressConsole plugin and their corresponding windows. (I am not sure if the current design of the console, especially the relation between main and slave window is a good one, either. What about interchanging their roles as in most cases users would expect the presentation view to respond to their actions inside the console window but not the other way round, even if this could sometimes be handy. But maybe we should discuss this separately.)

It would be good to have some kind of general plugin for that kind of synchronization, not only for media but  also for forms, hover/highlight/pointer events and so on, but I did not have an idea how to accomplish that without catching every single event by hand, filtering unwanted ones and propagating the interesting ones to the other window.

I would let it be the way it is at the moment.

Reply:
Maybe one could implement this console detection more generally in lib because it could be handy for some other plugins.
In this case the function should not trigger *any* actions.

Reply:
> That would mute media at the console view even when it is not played automatically.

Ah, good point. I haven't considered that use case.

You could still achieve the same goal while taking that into account, but I think my suggestion has in any case lost its elegance at that point and if anything your approach is now clearer.

> In this case the function should not trigger any actions.

I agree.